BLE_GAP/source/main.cpp

```diff
@@ -0,0 +1,494 @@
+        uint16_t timeout = scanning_params[_set_index].timeout;
+
+        /* active scanning will send a scan request to any scanable devices that
+         * we see advertising */
+        bool active = scanning_params[_set_index].active;
+
+        /* set the scanning parameters according to currently selected set */
+        error = _ble.gap().setScanParams(interval, window, timeout, active);
+
+        if (error) {
+            printf("Error during Gap::setScanParams\r\n");
+            return;
+        }
+
+        /* start scanning and attach a callback that will handle advertisements
+         * and scan requests responses */
+        error = _ble.gap().startScan(this, &GAPDevice::on_scan);
+
+        if (error) {
+            printf("Error during Gap::startScan\r\n");
+            return;
+        }
+
+        printf("Scanning started.\r\n");
+    };
+
+    /** After a set duration this cycles to the next demo mode
+     *  unless a connection happened first */
+    void on_duration_end()
+    {
+        print_performance();
+
+        /* alloted time has elapsed, move to next demo mode */
+        _event_queue.call(this, &GAPDevice::demo_mode_end);
+    };
+
+    /** Look at scan payload to find a peer device and connect to it */
+    void on_scan(const Gap::AdvertisementCallbackParams_t *params)
+    {
+        /* keep track of scan events for performance reporting */
+        _scan_count++;
+
+        /* don't bother with analysing scan result if we're already connecting */
+        if (_is_connecting) {
+            return;
+        }
+
+        /* parse the advertising payload, looking for a discoverable device */
+        for (uint8_t i = 0; i < params->advertisingDataLen; ++i) {
+            /* The advertising payload is a collection of key/value records where
+             * byte 0: length of the record excluding this byte
+             * byte 1: The key, it is the type of the data
+             * byte [2..N] The value. N is equal to byte0 - 1 */
+            const uint8_t record_length = params->advertisingData[i];
+            if (record_length == 0) {
+                continue;
+            }
+            const uint8_t type = params->advertisingData[i + 1];
+            const uint8_t* value = params->advertisingData + i + 2;
+
+            /* connect to a discoverable device */
+            if ((type == GapAdvertisingData::FLAGS)
+                && (*value & GapAdvertisingData::LE_GENERAL_DISCOVERABLE)) {
+
+                /* abort timeout as the mode will end on disconnection */
+                _event_queue.cancel(_on_duration_end_id);
+
+                printf("We found a connectable device\r\n");
+
+                ble_error_t error = _ble.gap().connect(
+                        params->peerAddr, Gap::ADDR_TYPE_RANDOM_STATIC,
+                        NULL,
+                        &connection_scan_params);
+
+                if (error) {
+                    printf("Error during Gap::connect\r\n");
```

Comment:
In case of error, the cycle can end immediately. With the current code it is in a state where `_on_duration_end_id` has been canceled and `_is_connecting` is marked as true.